Property implementation

paulidale opened the pull request:
A first pass at implementing some efficient property definition and query infrastructure.

Properties are defined by providers as strings which are parsed into name, value pairs.
Queries are parsed into name relation value triplets.
Implementations that match the query are considered for further use whereas implementations that don't match are immediately discarded from consideration.

The property system, allows global properties to be specified in addition to local properties (which can override globals).

If there is a match, there are capabilities for caching the outcome which will improve performance in the future.

The implementation database permits several interfaces to operate within:
1. add,
1. remove,
1. fetch,
1. cache_get,
1. cache_set and
1. set_global_properties

The generate usage pattern is:
```
if (cache_get(...) != NULL)
    return cache_got;

if ((result = fetch(...)) == NULL)
    return NULL;
cache_set(nid, properties, result)
```

- [x] tests are added or updated

levitte commented:
I added this PR to the newly create github project ["3.0 New Core + FIPS"](https://github.com/openssl/openssl/projects/2)

Also, you need to rebase.

levitte commented on crypto/property/defn_cache.c:
```diff
@@ -0,0 +1,101 @@
+#include <string.h>
+#include <openssl/err.h>
+#include <internal/propertyerr.h>
+#include <internal/property.h>
```
Internal inclusions are normally quoted.

(I've had a discussion elsewhere that lead me to believe that we really should do so with the openssl/ inclusions as well, since we *are* including project local headers rather than system headers, but that's a matter we'd need to discuss on the project list)

levitte commented:
I removed the FIPS label, as this isn't the FIPS module per se.  I believe adding this to the "3.0 New Core + FIPS" is a better way to mark these sorts of PRs

richsalz commented on crypto/property/defn_cache.c:
```diff
@@ -0,0 +1,101 @@
+typedef struct {
+    const char *prop;
+    PROPERTY_LIST *defn;
+    char body[1];
```
"unwarranted chumminess with the compiler" (web-search it)
Pay the price of portable code and make this a pointer.

paulidale replied:
Double allocating and ruining cache coherency in code that's meant to be fast to avoid something that's always worked...

richsalz replied:
I think that's a misguided/premature optimization, at the expense of more clear code. We're going to be dispatching to *crypto code* :)

t8m replied:
C99 allows using char body[]; Is C99 viable target for OpenSSL 3.0.0?

paulidale replied:
We can't use C99.  However, this code is portable and safe.  There are two restrictions: don't use an array of such structures (no space) and don't store a type other than char there (alignment).

mattcaswell commented on crypto/property/README:
```diff
@@ -0,0 +1,89 @@
+  names and values to small integer indices.  Names and values are
+  stored in separate hash tables.  The two Boolean values "yes" and "no"
+  are populated as the first two members of the value table.  The string
+  "name" is automatically populated as the first property name, to enable
```
Where does this happen? I couldn't see it.

paulidale replied:
See `ossl_property_parse_init` in `property_parse.c`.

mattcaswell replied:
That function only seems to define "provider", "version", "fips" and "engine", but not "name"

paulidale replied:
The question is which names are we going to reserve?

I changed "name" to "provider" since it seemed more applicable but missed the README.

paulidale replied:
The fix is to not say which property names are pre-loaded, just to say that they are.

paulidale replied:
We can have the discussion about which get added later.  I expect some worthwhile ones will _naturally_ appear over the course of the project.

richsalz commented on crypto/property/properties.ebnf:
```diff
@@ -0,0 +1,17 @@
+          ( ',' ( '-'? PropertyName | PropertyName ( '=' | '!=' ) Value ) )*
+Value    ::= NumberLiteral
+        | StringLiteral
+StringLiteral ::= QuotedString | UnquotedString
```
I think unquotedstring is a very big mistake.  It makes it really hard to add keywords to the query language in the future.  Look at perl's "barewords" evolution, for example.

richsalz commented on crypto/property/properties.ebnf:
```diff
@@ -0,0 +1,17 @@
+Value    ::= NumberLiteral
+        | StringLiteral
+StringLiteral ::= QuotedString | UnquotedString
+QuotedString     ::= '"' [^"]* '"'
```
I think you should allow backslash-escape inside the quoted string.

paulidale replied:
Why?  so we can write `fips="Y\e\s"` ?

I'd be happy to ban \ inside a string to provide a capability to add it painlessly in the future.

richsalz replied:
Banning it for future expansion helps.  But I meant to only escape quotes, as in things like "Hello \"world\""

levitte replied:
If anyone is ever going to want to have a Windows path in a quoted string, we'd better *not* use \ as the escape character.

That being said, the \ is more or less universally known as an escape character, and is already used as such elsewhere in OpenSSL, so it's actually a little surprising to have something different for property values.

(and I'm curious, where was ^ inspired from?  That's actually used in VMS file specs, where you'll see things like `openssl-1^.0^.2q^.tar.gz`, but I've not seen you as VMS savvy, so...)

mspncp replied:
> where was ^ inspired from?

The caret at the beginning of a character class inverts the set, i.e. `[^a-b]` is the complement of `[a-b]`. If you want to include a literal caret, it must not be first in the set. See GREP(1).

mspncp replied:
(_Of course_ you know that, it's part of your Perl-fu. It was just too early in the morning... ;-) ).

levitte replied:
D'oh.

And still, that was after the first coffee.

paulidale replied:
I think the ^ is stretching eBNF a little.

mattcaswell commented on test/property_test.c:
```diff
@@ -0,0 +1,369 @@
+    { "night", "day", "day=yes, night=yes" },
+    { "", "", "" },
+    { "-day", "day=yes", "day=no" },
+    { "-night, day", "day, night", "day=yes, night=no" },
```
I find these last two lines unexpected. The design doc has this example:
Global setting: fips=no
Passed setting: -fips
Resulting query: fips is not specified

Therefore I would have expected the above two lines to be:
````
    { "-day", "day=yes", "" },
    { "-night, day", "day, night", "day=yes" },
````

levitte replied:
I agree with @mattcaswell here

paulidale replied:
Is it the ordering of the fields that the problem?  They are ordered _local_, _global_ and _result_.  If I switch them for _global_, _local_ then _result_, it would be okay?  Did the `"colour=blue", "colour=red", "colour=blue"` earlier confuse too?

Or is it the =no tests (meaning test for undefined or "no")?

slontis replied:
Yep the ordering is confusing, Switch them and maybe change the struct names to make it more obvious.

paulidale replied:
Ordering switched and better names allocated.

levitte replied:
For me, it was that the combo of "-foo" and "foo=yes" would give "foo=no". As far as I understood, the result should be nothing, i.e. "foo" unspecified.

paulidale replied:
We can't query for "unspecified", just "yes" or "no".  "foo=no" is true if unspecified however (both being false).

I think this is the corrent behaviour, since no provider will set everything it doesn't do to "no" but it will set what it does do to "yes".

paulidale replied:
> { "-day", "day=yes", "" },

Noting that the empty query here would match everything regardless of what is set and what isn't.  This isn't what needs to be validated as working at this point.

levitte replied:
Uhmmm, so here's the examples in the design document, under "Multileve Queries":

-   global: fips=yes
    passed: -fips
    resulting: fips is not specified
-   global: fips=no
    passed: -fips
    resulting: fips is not specified

Could you tell us how that corresponds to what you just said?

mattcaswell replied:
Hmmm. To me "fips=no" means do not select an algorithm defined as "fips=yes". Whereas "" means I don't care whether you select an algorithm with "fips=yes" defined or not.

mattcaswell replied:
...and so of course global "fips=yes" and passed "-fips" means I don't care whether you select fips or not - which is not the same as "fips=no".

paulidale replied:
I think the disconnect here is that these are _test_ cases, not _examples_ of usage.

How about I add another case "fips=yes" globally, "-fips" locally and "fips=yes" as the definition?  Or "fips=yes" then "-fips" and a definition of "sesquioxidizing=1960"?

paulidale replied:
I've added some additional test cases for the '-' prefix.

mattcaswell replied:
This helps immensely! It suddenly all makes sense to me. Actually I think the source of the confusion (in part) is the function name `ossl_property_compare`. One intuitively expects that function to take two property lists and return true if they are equal. Hence merging local "-day" with global "day=yes" and saying that is equal to "day=no" is not what I expected. Actually it doesn't "compare" lists exactly at all. It tells you whether a query matches against a property definition. Perhaps a different name for the function?

mattcaswell commented on test/property_test.c:
```diff
@@ -0,0 +1,369 @@
+    int i, res = 0;
+    char buf[50];
+    void *result;
+    int err = 0;
```
Having a variable name the same as a label is confusing.

levitte commented on crypto/property/defn_cache.c:
```diff
@@ -0,0 +1,101 @@
+
+DEFINE_LHASH_OF(PROPERTY_DEFN_ELEM);
+
+static LHASH_OF(PROPERTY_DEFN_ELEM) *property_defns = NULL;
```
I'm about to submit that PR.  It is fairly standalone :-)

richsalz replied:
Initializing static variables to NULL is not common.

levitte replied:
What would you initialie it to, @richsalz ?

richsalz replied:
I would remove the ```= NULL```  For example, your context PR doesn't do this.

levitte replied:
1. The default context isn't a pointer.
2. The default context is initialized on demand through a RUN_ONCE function, so what you imply is that the property stuff use a similar mechanism to get initialized.

Point is, this *must* be properly initialized *somehow*.

richsalz replied:
Omit the NULL assignment and the pointer will still be properly initialized.

t-j-h replied:
Init of statics like this is a reasonable style. I wouldn't change it.

paulidale replied:
This one wouldn't need to be in the library context.  It is a space saving cache and can be safely shared among all contexts.

paulidale replied:
I've gone and removed the global store and will rely on the library context to provide a reference.

levitte replied:
> and will rely on the library context to provide a reference.

Actually, you still have to provide some code to ask for a reference.  The library context is just a bag of stuff that the context itself knows absolutely nothing about, everyone who wants in must ask for a spot (an index, really).  I did document it :wink:

paulidale replied:
The race to see which PR is merged first is on :)

levitte replied:
I win

levitte commented on crypto/property/property.c:
```diff
@@ -0,0 +1,507 @@
+
+DEFINE_SPARSE_ARRAY_OF(ALGORITHM);
+
+static OSSL_IMPL_STORE *g_implementations = NULL;
```
We will need a "global" store (in the library context when that shows up)

paulidale replied:
I concur, this one should be in the library context not as a global.

levitte commented on crypto/property/property.c:
```diff
@@ -0,0 +1,507 @@
+}
+
+int ossl_impl_store_set_global_properties(OSSL_IMPL_STORE *impls,
+                                          const char *prop_query) {
```
It does for everything that will use this store, and that goes for everything EVP

paulidale replied:
Could we call them "store" properties instead of "global"?

paulidale replied:
Or "context" properties?

levitte commented on include/internal/property.h:
```diff
@@ -0,0 +1,39 @@
+                           int nid, const void *implementation);
+int ossl_impl_store_fetch(OSSL_IMPL_STORE *cache,
+                          int nid, const char *properties,
+                          const void **result);
```
Please make that non-void.  There's a good chance there will be refcounted objects...

levitte replied:
Oh wait, hold on, I might be wrong.  Deleting this if I can

levitte commented on include/internal/property.h:
```diff
@@ -0,0 +1,39 @@
+void ossl_impl_store_free(OSSL_IMPL_STORE *cache);
+int ossl_impl_store_init(void);
+void ossl_impl_store_cleanup(void);
+int ossl_impl_store_add(OSSL_IMPL_STORE *cache,
```
s/cache/store/

levitte commented on include/internal/property.h:
```diff
@@ -0,0 +1,39 @@
+
+/* Implementation store functions */
+OSSL_IMPL_STORE *ossl_impl_store_new(void);
+void ossl_impl_store_free(OSSL_IMPL_STORE *cache);
```
s/cache/store/

levitte commented on include/internal/property.h:
```diff
@@ -0,0 +1,39 @@
+                        int nid, const char *properties,
+                        void *implementation,
+                        void (*implementation_destruct)(void *));
+int ossl_impl_store_remove(OSSL_IMPL_STORE *cache,
```
s/cache/store/

levitte commented on include/internal/property.h:
```diff
@@ -0,0 +1,39 @@
+                        void (*implementation_destruct)(void *));
+int ossl_impl_store_remove(OSSL_IMPL_STORE *cache,
+                           int nid, const void *implementation);
+int ossl_impl_store_fetch(OSSL_IMPL_STORE *cache,
```
s/cache/store/

levitte commented on include/internal/property.h:
```diff
@@ -0,0 +1,39 @@
+int ossl_impl_store_fetch(OSSL_IMPL_STORE *cache,
+                          int nid, const char *properties,
+                          const void **result);
+int ossl_impl_store_set_global_properties(OSSL_IMPL_STORE *impls,
```
s/impls/store/

paulidale replied:
done throughout propery.c and property.h

levitte commented on include/internal/property.h:
```diff
@@ -0,0 +1,39 @@
+                                          const char *prop_query);
+
+/* proeprty query cache functions */
+int ossl_impl_store_cache_get(OSSL_IMPL_STORE *c, int nid, const char *prop,
```
Let's not have single-letter parameters...

levitte replied:
Also, s/prop/propquery/

paulidale replied:
Done.

levitte commented on include/internal/property.h:
```diff
@@ -0,0 +1,39 @@
+/* proeprty query cache functions */
+int ossl_impl_store_cache_get(OSSL_IMPL_STORE *c, int nid, const char *prop,
+                              void **result);
+int ossl_impl_store_cache_set(OSSL_IMPL_STORE *c, int nid, const char *prop,
```
Same as above

levitte commented:
Looking through the code again and its intended use, I'm unsure that the functions declared in `include/internal/property.h` are named right.

That store (or those stores) are going to be used to save away methods, i.e. instances of EVP_CIPHER, EVP_MD, etc.  Do we call them implementations or do we call them methods?  Historically, they were called the latter (it's even more visible if you consider EVP_PKEY_METHOD), so I wonder of `ossl_meth_` isn't a better function prefix than `ossl_impl_`.

Food for thought.

levitte commented on include/internal/property.h:
```diff
@@ -0,0 +1,39 @@
+                           int nid, const void *implementation);
+int ossl_impl_store_fetch(OSSL_IMPL_STORE *cache,
+                          int nid, const char *properties,
+                          const void **result);
```
Please make that non-const.  The result may be a refcounted object (I actually expect them to be) that will need an update, plus it's consistent with `ossl_impl_store_cache_get` and `ossl_impl_store_cache_set`

paulidale replied:
Done.

levitte commented on test/build.info:
```diff
@@ -484,6 +484,10 @@ IF[{- !$disabled{tests} -}]
     DEPEND[wpackettest]=../libcrypto ../libssl.a libtestutil.a
 
+    SOURCE[property_test]=property_test.c
+    INCLUDE[property_test]=../include
```
You need to add `../apps/include` here

levitte commented on crypto/property/property.c:
```diff
@@ -0,0 +1,507 @@
+
+    ossl_impl_store_cleanup();
+    return 0;
+}
```
Please make sure that this is called automatically, either as part of the init process, or in here with the help of a RUN_ONCE function (see #8225 to see how I do that)

levitte commented on crypto/property/property.c:
```diff
@@ -0,0 +1,507 @@
+    ossl_prop_defn_cleanup();
+    ossl_impl_store_free(g_implementations);
+    g_implementations = NULL;
+}
```
Please make sure that this is called automatically, either as part of the cleanup process, or in here with the help of `OPENSSL_atexit()` (see #8225 to see how I do that)

levitte commented on crypto/property/property.c:
```diff
@@ -0,0 +1,507 @@
+    int ret = 0;
+
+    if (impls == NULL)
+        return 1;
```
Why not in `g_implementations`?

paulidale replied:
It should be there, but as mentioned earlier, I suspect it is better if `g_implementations` disappears completely.

paulidale replied:
I've removed `g_implementations` entirely.  The library context will need to provide.

paulidale commented on crypto/property/README:
```diff
@@ -0,0 +1,89 @@
+
+* A cache for property definitions is also viable, if only implementation
+  properties are used and not algorithm properties, or at least these are
+  maintained separately.  This cache would be an hash table, indexed by
```
I say the h quite softly.  Almost but not quite 'ash'.  I'm not fussed and would be happy to go either way.

paulidale commented on crypto/property/defn_cache.c:
```diff
@@ -0,0 +1,101 @@
+        p->defn = pl;
+        memcpy(p->body, prop, len + 1);
+        if (lh_PROPERTY_DEFN_ELEM_insert(property_defns, p) != NULL
+            || lh_PROPERTY_DEFN_ELEM_retrieve(property_defns, &elem) != NULL)
```
Yes, I should and have now.

paulidale commented on crypto/property/defn_cache.c:
```diff
@@ -0,0 +1,101 @@
+    return r != NULL ? r->defn : NULL;
+}
+
+int ossl_prop_defn_set(const char *prop, PROPERTY_LIST *pl)
```
It is setting a mapping from the string to the compiled list.  I.e. array[prop] = pl.  We normally put the assigned item on the left don't we?

I'm happy to switch the argument around but they make sense to me this way.

t-j-h replied:
Switching args around should be done - our convention is to have the function prefix match the first parameter in the argument list in terms of naming (generally).

paulidale commented on crypto/property/property.c:
```diff
@@ -0,0 +1,507 @@
+        impls = g_implementations;
+
+    /* Only a read lock because no attempt is made to clean up empty stacks */
+    ossl_property_write_lock(impls);
```
Comment was old.  I toyed with a two level locking scheme but it was overly complex so I removed the lower level.

paulidale commented on crypto/property/property.c:
```diff
@@ -0,0 +1,507 @@
+        p2 = ossl_property_merge(pq, impls->global_properties);
+        if (p2 == NULL)
+            goto fin;
+        pq = p2;
```
Yes :(

paulidale commented on crypto/property/property.c:
```diff
@@ -0,0 +1,507 @@
+    unsigned int n;
+
+    if (impls->nbits == 0) {
+        if (!RAND_bytes(impls->rand_bits, sizeof(impls->rand_bits)))
```
You managed to figure it out, so ideal code :)
I'll add a comment.

paulidale commented on crypto/property/property_parse.c:
```diff
@@ -0,0 +1,541 @@
+        return 0;
+    do
+        v = v * 10 + (*s++ - '0');
+    while (ossl_isdigit(*s));
```
I've added them throughout.

paulidale commented on crypto/property/README:
```diff
@@ -0,0 +1,89 @@
+  "name" is automatically populated as the first property name, to enable
+  algorithm selection via properties as well as by NID.  No functions
+  are provided to convert from an index back to the original string
+  (this can be done by maintaining parallel stacks of strings).
```
Yes, old comment.  The question is do we want the ability to convert backwards?  There currently is no export of anything that would grant access to the indicies and no use of them.  I can only see them being useful as a diagnostic aid.

I suggest we make the comment correct and remove the relevant code.  Thoughts?

paulidale replied:
I've removed the reverse mappings.

paulidale commented on crypto/property/defn_cache.c:
```diff
@@ -0,0 +1,101 @@
+        p->prop = p->body;
+        p->defn = pl;
+        memcpy(p->body, prop, len + 1);
+        if (lh_PROPERTY_DEFN_ELEM_insert(property_defns, p) != NULL
```
Yes, it was.  And on cache flush.  Fixed now.

paulidale commented:
I've addressed most of the comments (resolving those that I did fix).  This mostly leaves the library context (& `g_implementations`) some parsing questions and the run once.

paulidale commented:
> Looking through the code again and its intended use, I'm unsure that the functions declared in `include/internal/property.h` are named right.

Let's settle this before I do the documentation.

paulidale commented:
Travis failures look to be overly aggressive detection of unused macro generated functions.  Do we have a standard way of prevent these without exposing the structures via a header file?

slontis commented on crypto/property/README:
```diff
@@ -0,0 +1,88 @@
+Some consequences of these assumptions are:
+
+* That definition is uncommon and queries are very common, we can treat
+  the property definitions are most immutable.  Specifically, a query can
```
nit - bad sentence

slontis commented on crypto/property/README:
```diff
@@ -0,0 +1,88 @@
+
+  The caching of query results has to be efficient but it must also be robust
+  against a denial of service attack.  The cache cannot be permitted to grow
+  withut bounds and must garbage collect under used entries.  The garbage
```
nit: withut

paulidale commented:
Items left include:
1. function names
1. documentation
1. automatic initialisation
1. unquoted strings
1. escapes in strings
1. chumminess

levitte commented:
> Travis failures look to be overly aggressive detection of unused macro generated functions. Do we have a standard way of prevent these without exposing the structures via a header file?

I agree this is annoying.  I assume it's a warning flag that comes with -Wall or -Wextra.  A bit of experimenting shows it's clang being more strict about `-Wunused-function` than gcc is.  And yes, clang is quite aggressive.  The way to turn this off is `-Wno-unused-function`.

The benefit with `-Wunused-function` is that it helps us detect dead code.  But considering what silliness it pushes us to (such as *forcing* us to place DEFINE macro calls in headers), I'm unsure that the benefit is worth it...

paulidale commented:
In this case it forces the typedefs of file private types into a header.  Sigh.
Anyway, I expand all the functions in the local header now to avoid the issue.
I presume that this means some compilers will generate many copies of the functions as well.

levitte commented:
Well, we can disable that warning for clang specifically, leaving it to gcc to still detect unused functions, just a bit more kindly.

levitte commented:
@paulidale, #8234 solves the problem, I think...

paulidale commented:
It looks like it should work.  I can revert the commit moving everything tomorrow and try.

levitte commented:
I found that #8234 wasn't enough, but I believe #8235 completes it.

levitte commented on test/property_test.c:
```diff
@@ -129,6 +129,9 @@ static const struct {
     { "day=yes", "-day", "day=no" },
+    { "day=yes", "-day", "day=yes" },
+    { "day=yes", "-day", "day=arglebargle" },
+    { "day=yes", "-day", "pot=sesquioxidizing" },
```
Just to ensure that the likes of @mattcaswell and me don't start asking questions again in a year or two, it might be good to add a comment above `{ "day=yes", "-day", "day=no" },`, saying something like:

``` C
    /* The following four leave 'day' unspecified in the query, and will match any definition */

paulidale replied:
Done.

levitte commented on crypto/property/property.c:
```diff
@@ -0,0 +1,496 @@
+    }
+}
+
+OSSL_METHOD_STORE *ossl_method_store_new(void)
```
function rename noticed!  [nods approvingly]

levitte commented on crypto/property/property_string.c:
```diff
@@ -0,0 +1,164 @@
+ * space and time efficient algorithms if this becomes a bottleneck.
+ */
+
+typedef struct {
```
Unnamed struct might be uncommon, but naming it becomes a bit silly if that name is never used.  I see no reason to add a struct name now.

levitte commented on doc/internal/man3/OSSL_METHOD_STORE.pod:
```diff
@@ -0,0 +1,101 @@
+
+ossl_method_store_add() adds the B<implementation> to the B<store> as an
+instance of the algorithm indicated by B<nid> and the property definition
+<properties>. The optional B<implementation_destruct> function is called when
```
May I suggest breaking the line after each end of sentence (i.e. the full stop)?  This is to avoid the dispute of one vs two spaces after a full stop, and lets local conventions decide instead (it has been observed that different installations have different conventions, but those will only be applied on a linebreak within a paragraph)

levitte replied:
I try to apply what I just wrote on any *new* POD file I write (or old ones if I rewrite them entirely).  Doing this on all existing POD files is a different story.

paulidale replied:
Done.

levitte commented on include/openssl/lhash.h:
```diff
@@ -131,6 +132,10 @@ void OPENSSL_LH_node_usage_stats_bio(const OPENSSL_LHASH *lh, BIO *out);
     { \
         OPENSSL_LH_free((OPENSSL_LHASH *)lh); \
     } \
+    static ossl_inline void lh_##type##_flush(LHASH_OF(type) *lh) \
```
Depending on what happens with #8246, you may have to add a `ossl_unused` here

levitte commented on include/openssl/safestack.h:
```diff
@@ -48,6 +48,10 @@ extern "C" {
     { \
         return OPENSSL_sk_reserve((OPENSSL_STACK *)sk, n); \
     } \
+    static ossl_inline int sk_##t1##_reserve_at_least(STACK_OF(t1) *sk, int n) \
```
... and here

paulidale replied:
Once the reverse property string lookup was removed, this isn't required anymore.  It was a bit of a kludge so better without it.

levitte commented:
> I found that #8234 wasn't enough, but I believe #8235 completes it.

The final solution for the DEFINE issue was #8246

levitte commented:
@paulidale, for automatic init, I propose 31bab9279980df5fd1f952eb02ca49a490184ff4 (from https://github.com/levitte/openssl/tree/300-property%2BRL)

levitte commented:
Hey, I noticed 412da9ecdfca23b6c3d964dd9332d0ef90ef58c2, and just wanted to say that I think it's a good move (and it's easy to deal with)

richsalz commented:
*   If anyone is ever going to want to have a Windows path in a quoted string, we'd better not use \ as the escape character.

Some options come to mind; allow \\ to mean a single \, or users can use /



.

paulidale commented:
> @paulidale, for automatic init, I propose [31bab92](https://github.com/openssl/openssl/commit/31bab9279980df5fd1f952eb02ca49a490184ff4) (from https://github.com/levitte/openssl/tree/300-property%2BRL)

I've no objection to this.  I do init and cleanup multiple times in the tests which might confound things.

paulidale commented:
Squashed and new changes added.  The initialisation code causes test problems (more on this later).

I think we're down to allowing unquoted strings and to allow escapes or not in quoted strings.

For unquoted strings, I'm not too concerned about running into reserved keywords later.  The context is sufficient to resolve any issues.  The string "name=" **_must_** be followed by an unquoted string.  The string "name" **_must not_** be followed by an unquoted string.

Unquoted strings also permit case insensitive values.  Character set issues prevent quoted strings from having their case massaged.  This might have some potential for user confusion.

If we ever switch to using a parser generator, keywords might introduce lexical analyser pain.  FORTRAN avoided reserved words in the 1950's as have other languages since.


Escapes in quoted strings is a harder.  I'm having trouble thinking of a reasonable case where they are required but users will find a way.  As it currently stands, either **'** or **"** can be used to quote strings.  If your string contains _"_, use **'** and if it contains _'_, use **"**.  The only difficulty arises if the value string contains both _'_ and _"_.  I can live with this restriction.

Adding _\\_ escapes to strings introduces complexity:

- Window's paths or path conversion and
- numeric characters (decimal & octal to match C and quite possibly hexadecimal as well).  I simply don't want to go here.  Supporting octal numbers added a surprising amount of complexity to the grammar.  How many people these days use octal constants when writing C?  How often is octal used in OpenSSL properties?  Currently, only the test case to see that they work 😉

paulidale commented:
> for automatic init, I propose [31bab92](https://github.com/openssl/openssl/commit/31bab9279980df5fd1f952eb02ca49a490184ff4)

This breaks the (current) unit tests.  The property subsystem is re-initialised a couple of times during their execution.

I was looking at introducing initialisation via `crypto/init.c` but it wasn't clear how to best do it.  That it doesn't have a user facing API seems to be part of the cause of my doubt because everything else there does.

I guess it comes down to getting the unit tests going or finding another way.  I'm open to suggestions as to which is preferable.

levitte commented:
Ok...  Just now, I noticed that you do declare `ossl_method_store_init` and `ossl_method_store_cleanup`, so you seem to expect someone else to do the initialization call, am I right?  Trouble is, those are a global one-time shot.  Meanwhile, you provide the opportunity for anyone to create a new store at any time, which assumes that `ossl_method_store_init` has been called first.  

The only issue I see with that approach is that I can't see a way for other code to determine if `ossl_method_store_init` has already been called or not.  That's a recipe for fragility.

paulidale commented:
It will require a RUN_ONCE somewhere.  Why not `OPENSSL_init_crypto` for it?

levitte commented:
> Why not `OPENSSL_init_crypto` for it?

Personal preference.  I much prefer having things together, so it's immediately visible what is a once-in-a-process-life init and when it's called.  I also tend to like the on-demand paradigm that follows with the local RUN_ONCE calls.
I would actually prefer if only a minimum amount of stuff was done directly in `crypto/init.c`, i.e. the very few stuff that really everything else depends on and must be taken down in a deterministic order.

But either way, whether we do my RUN_ONCE dance or have the init code called from `crypto/init.c`, you still have the exact same reinitialization problem.

paulidale commented:
I've have a plan....

paulidale commented:
and the plan needs valgrind to continue.

FdaSilvaYY commented on crypto/property/defn_cache.c:
```diff
@@ -0,0 +1,105 @@
+#include <openssl/err.h>
+#include "internal/propertyerr.h"
+#include "internal/property.h"
+#include <openssl/lhash.h>
```
nit : include order is a bit messy ?

paulidale replied:
Fixed.

paulidale commented on crypto/property/README:
```diff
@@ -0,0 +1,87 @@
+
+  The caching of query results has to be efficient but it must also be robust
+  against a denial of service attack.  The cache cannot be permitted to grow
+  without bounds and must garbage collect under used entries.  The garbage
```
I've hyphenated it.

paulidale commented on crypto/property/property.c:
```diff
@@ -0,0 +1,508 @@
+ * In order to avoid taking a write lock to keep accurate LRU information or
+ * using atomic operations to approximate similar, the procedure used here
+ * is to stochastically flush approximately half the cache.  Since generating
+ * random numbers is relatively expensive, we proiduce them in blocks and
```
Done

paulidale commented:
Tests are updated so that they don't reinitialise the properties anymore.
I think that's everything once the CIs pass.

paulidale commented:
The CIs pass :)

paulidale commented:
Which PROPERTY ?  Everything exposed outside of the crypto/property starts OSSL_ doesn't it?

t-j-h commented:
You have ossl_ prefixes on the functions expected to be used outside of the property directory but no prefix on the typedefs. They should be OSSL_ ... basically we don't want to run into clashes with other code and PS_IDX and PROPERTY are likely to clash. 

Yes I know these aren't currently expected to be visible elsewhere - but they may evolve that way - and we should be clean in our "namespace" usage IMHO.

paulidale commented:
I've prefixed the internal symbols.

paulidale commented:
@levitte @mattcaswell I'll merge this in the morning (my time) unless there are objections.

paulidale commented:
Squashed...

levitte commented:
I'd like to run a test tomorrow against things I have that build on this...

levitte commented:
(It looked good visually, though...)

paulidale commented:
Let me know when you're done and I'll merge it or merge it yourself if the time-zones are immiscible.

paulidale commented:
Merged, thanks for the extensive and positive feedback.

levitte commented:
... someone's in a rush.  :wink:

My tests ran well after some appropriate mods.  The manual bits need some fixups, something I discovered moments ago...

paulidale commented:
In a rush or needing a rush?  I waited until I thought I’d be asleep (which I have been).  That I’m not indicative is indactive of a non-problem….


Pauli
-- 
Dr Paul Dale | Cryptographer | Network Security & Encryption 
Phone +61 7 3031 7217
Oracle Australia



> On 18 Feb 2019, at 6:00 pm, Richard Levitte <notifications@github.com> wrote:
> 
> ... someone's in a rush. 😉
> 
> My tests ran well after some appropriate mods. The manual bits need some fixups, something I discovered moments ago...
> 
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub <https://github.com/openssl/openssl/pull/8224#issuecomment-464625182>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ARFiwdynDshahU7-Wch_GVzyCWHeZRDOks5vOl2IgaJpZM4a5LHK>.
>

levitte commented:
Care to look at #8265?
